src/Plugins/ImageProcessor.Web/ImageProcessor.Web.Plugins.AmazonS3Cache/AmazonS3Cache.cs

```diff
@@ -0,0 +1,493 @@
+
+                    if (response != null)
+                    {
+                        cachedImage = new CachedImage
+                        {
+                            Key = key,
+                            Path = this.CachedPath,
+                            CreationTimeUtc = response.LastModified
+                                                      .ToUniversalTime()
+                        };
+
+                        CacheIndexer.Add(cachedImage);
+                    }
+                }
+                catch (Amazon.S3.AmazonS3Exception ex)
+                {
+                    // Nothing in S3 so we should return true.
+                    isUpdated = true;
+                }
+            }
+
+            if (cachedImage == null)
+            {
+                // Nothing in the cache so we should return true.
+                isUpdated = true;
+            }
+            else
+            {
+                // Check to see if the cached image is set to expire.
+                if (this.IsExpired(cachedImage.CreationTimeUtc))
+                {
+                    CacheIndexer.Remove(this.CachedPath);
+                    isUpdated = true;
+                }
+            }
+
+            return isUpdated;
+        }
+
+        /// <summary>
+        /// Adds the image to the cache in an asynchronous manner.
+        /// </summary>
+        /// <param name="stream">
+        /// The stream containing the image data.
+        /// </param>
+        /// <param name="contentType">
+        /// The content type of the image.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Task"/> representing an asynchronous operation.
+        /// </returns>
+        public override async Task AddImageToCacheAsync(Stream stream, string contentType)
+        {
+            var transferUtility = new TransferUtility(this.amazonS3ClientCache);
+
+            var path = GetFolderStructureForAmazon(this.CachedPath);
+            var filename = Path.GetFileName(this.CachedPath);
+            var key = GetKey(path, filename);
+
+            var transferUtilityUploadRequest = new TransferUtilityUploadRequest
+            {
+                BucketName = this.AwsBucketName,
+                InputStream = stream,
+                Key = key,
+                CannedACL = S3CannedACL.PublicRead
+            };
+
+            await transferUtility.UploadAsync(transferUtilityUploadRequest);
+        }
+
+        /// <summary>
+        /// Trims the cache of any expired items in an asynchronous manner.
+        /// </summary>
+        /// <returns>
+        /// The asynchronous <see cref="Task"/> representing an asynchronous operation.
+        /// </returns>
+        public override async Task TrimCacheAsync()
+        {
+            // At the present time ListObjectsRequest doesn't work.
+            // TODO: find an alternative solution? 
+            return;
```

Comment:
I think there might be an issue with the currect AWSSDK version (2.3.25): ListObjectsRequest was working fine in previous versions (I've tested it in 1.5.18)